api/v1/harborcluster_types.go

```diff
@@ -16,25 +16,321 @@ limitations under the License.
+	Accountkey  string `json:"accountkey,omitempty"`
+	Container   string `json:"container,omitempty"`
+	Realm       string `json:"realm,omitempty"`
+}
+
+// InClusterOptions of storage.
+type InClusterOptions struct {
+	// inCluster Provider, just support minIO now.
+	Provider string     `json:"provider,omitempty"`
+	Spec     *MinIOSpec `json:"spec,omitempty"`
+}
+
+type MinIOSpec struct {
+	// Supply number of replicas.
+	// For standalone mode, supply 1. For distributed mode, supply 4 or more (should be even).
+	// Note that the operator does not support upgrading from standalone to distributed mode.
+	Replicas int `json:"replicas,omitempty"`
+	// Version defines the MinIO Client (mc) Docker image version.
+	Version string `json:"version,omitempty"`
+	// VolumeClaimTemplate allows a user to specify how volumes inside a MinIOInstance
+	// +optional
+	VolumeClaimTemplate corev1.PersistentVolumeClaim `json:"volumeClaimTemplate,omitempty"`
+	// If provided, use these requests and limit for cpu/memory resource allocation
+	// +optional
+	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+type Database struct {
+	// Set the kind of which redis service to be used, inCluster or external.
+	Kind string `json:"kind,omitempty"`
+
+	// inCluster params following.
+	Storage          string              `json:"storage,omitempty"`
+	Replicas         int                 `json:"replicas,omitempty"`
+	Version          string              `json:"version,omitempty"`
+	StorageClassName string              `json:"storageClassName,omitempty"`
+	Resources        corev1.ResourceList `json:"resources,omitempty"`
+
+	// External params following.
+	// The secret must contains "address:port","usernane" and "password".
+	SecretName     string `json:"secretName,omitempty"`
+	SslConfig      string `json:"sslConfig,omitempty"`
+	ConnectTimeout int    `json:"connectTimeout,omitempty"`
+}
+
+type Redis struct {
+	// Set the kind of which redis service to be used, inCluster or external. Setting up a harbor-cluster with external redis service should provide client params to communicate. The difference between inCluster redis and external redis is that the inCluster redis installed automatically.
+	Kind string `json:"kind,omitempty"`
+
+	// inCluster params following.
+	Server   *RedisServer `json:"server,omitempty"`
+	Sentinel *Sentinel    `json:"sentinel,omitempty"`
+
+	// External params following.
+	// The secret must contains "address:port","usernane" and "password".
+	SecretName string `json:"secretName,omitempty"`
+	// Maximum number of socket connections.
+	// Default is 10 connections per every CPU as reported by runtime.NumCPU.
+	PoolSize int `json:"poolSize,omitempty"`
+	// TLS Config to use. When set TLS will be negotiated.
+	// set the secret which type of Opaque, and contains "tls.key","tls.crt","ca.crt".
+	// optional
+	TlsConfig string `json:"tlsConfig,omitempty"`
+}
+
+type Sentinel struct {
+	Replicas int `json:"replicas,omitempty"`
```

Comment:
Sentinel also need a monitor name used to connect

Reply:
```yaml
redis:
  kind: external
  spec:
      groupName: mygroup # optional
      schema: sentinel # schema: redis
      secret: mySecret  #  --password=pass
      hosts:
        - host: locahost1
           port: 26379
        - host: localhost2
           port: 26379
 ```